### ISSUE TITLE
Kira for audio in bevy

### ISSUE THREAD

**Issue by Dash-L** [bevyengine/bevy#1304]
A while ago I opened an issue about how audio in its current state in `bevy` is very lacking in features. At the time I wanted to try to basically implement most of `rodio`'s `Source` trait and make that into `bevy`'s audio "API". However, I have recently come across a crate called [kira](https://github.com/tesselode/kira) which is an audio crate specifically targeted at games. What do you guys think about switching `rodio` with `kira`? It looks like at the moment games could probably get away with using `bevy` and `kira` separately, but I understand `bevy` likes having things all in one. If I were to try to improve bevy's audio API, should I stick to `rodio` or use `kira` instead? Or should audio be like networking right now and you just bring your own external crate?

**Comment by Ratysz** [comment]
Also consider: [`oddio`](https://github.com/Ralith/oddio). Same author as `hecs`, on which `bevy_ecs` is based on.

**Comment by cart** [comment]
Kira is a very promising project thats definitely worth considering for adoption. @Moxinilian has been investigating this, so I'll defer to them for status / opinions.

**Comment by Moxinilian** [comment]
I’ve been working closely with tesselode (the author of Kira) in the past few month to understand and tweak Kira towards better integration within Bevy (notably for it to support wasm and arbitrary audio sources). One obvious and great advantage of using Kira directly is that it has and will have many useful features tailored for video games, which lets us not have to maintain that (at least directly). I have not yet had time to look into oddio however.

Regarding integration itself, so far I have found an elegant way to do it would be using atelier-assets’ abstraction, so we could turn most Kira constructs into assets. If atelier-assets integration takes too much time however, it is possible I think to create an API that will basically copy ECS state into the Kira sound space via a system. There are still a lot of implementation details to consider (notably on how to take advantage of the parallel API for Kira that tesselode recently added). But something can definitely be done and I believe Kira to be of sufficient quality for integration.

**Comment by Dash-L** [comment]
Sounds neat! Unfortunately, this sounds like it's a little past my skill level/understanding but I can't wait to see how audio in bevy improves and is implemented.

**Comment by NiklasEi** [comment]
I started working on an audio plugin using Kira here: https://github.com/NiklasEi/bevy_improved_audio
A **very** basic implementation is done and works (only with `mp3` though). I am quite new to rust/bevy and will take any help I can get.

> Regarding integration itself, so far I have found an elegant way to do it would be using atelier-assets’ abstraction, so we could turn most Kira constructs into assets. If atelier-assets integration takes too much time however, it is possible I think to create an API that will basically copy ECS state into the Kira sound space via a system.

@Moxinilian could you maybe go into a bit more detail for me with these ideas? For now I've been orienting myself at the current `bevy_audio` implementation and would just try to expose more functionality like pausing/looping and so on.

**Comment by targrub** [comment]
Obviously, this work is farther along than this issue indicates.  Should this be closed?
